### PR TITLE
[WIP] Migrate Host -> Service Shared GW to OpenFlow

### PR DESCRIPTION
**Pull request by trozet**
This will allow Host access to services without DGP.

Signed-off-by: Tim Rozet <trozet@redhat.com>

### CODE REVIEW

**Comment by trozet** [comment]
need to fix unit tests

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 1]
```diff
@@ -231,6 +232,29 @@ func newSharedGatewayOpenFlowManager(patchPort, macAddress, gwBridge, gwIntf str
...
+			fmt.Sprintf("cookie=%s, priority=500, in_port=%s, ip, ip_dst=%s, ip_src=%s,"+
+				"actions=ct(commit,zone=%d,nat(dst=%s),table=3)",
+				defaultOpenFlowCookie, ofportPatch, types.V4HostMasqueradeIP, config.Gateway.V4JoinSubnet,
+				secondaryCTZone, physicalIP.IP))
+
+		// table 0, Reply SVC traffic from Host -> OVN, unDNAT and goto table 2
+		dftFlows = append(dftFlows,
+			fmt.Sprintf("cookie=%s, priority=500, in_port=LOCAL, ip, ip_dst=%s,"+
+				"actions=ct(zone=%d,nat,table=2)",
+				defaultOpenFlowCookie, config.Gateway.V4JoinSubnet, secondaryCTZone))
```
Comment (zshi-redhat) [inline go-controller/pkg/node/gateway_shared_intf.go:252]: why ip_dst is not matching V4HostMasqueradeIP given this is not for hairpin traffic?

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 2]
```diff
@@ -291,6 +378,37 @@ func newSharedGatewayOpenFlowManager(patchPort, macAddress, gwBridge, gwIntf str
...
+	dftFlows = append(dftFlows,
+		fmt.Sprintf("cookie=%s, table=3, "+
+			"actions=move:NXM_OF_ETH_DST[]->NXM_OF_ETH_SRC[],mod_dl_dst=%s,output:LOCAL",
+			defaultOpenFlowCookie, macAddress))
+
+	// table 4, hairpinned pkts that need to go from OVN -> external.
+	// We need to ct default zone so we can distinguish return traffic.
+	if config.IPv4Mode {
+		dftFlows = append(dftFlows,
+			fmt.Sprintf("cookie=%s, table=4,ip, "+
```
[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/node/gateway_shared_intf.go:397]: It doesn't. Reply traffic from external -> OVN for this case will first go to table 1 and be CT'ed into zone 64000, if we find a match there we know this is return traffic and not new traffic, so then we attempt to unDNAT it it in zone0:
https://github.com/ovn-org/ovn-kubernetes/pull/1984/commits/676bc601960ed426b07f02ecfed475f2b90ac9d4#diff-d3aa58d9b58a0a09264f072df46ab01d0501eb508c4656411ae2dc1ac68fb3c4R325

and then send it into table 5 to be dispatched into OVN.

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 1 above)
Comment (trozet) [reply go-controller/pkg/node/gateway_shared_intf.go:252]: This is for reply traffic that will come from the host back to ovn from a hairpin. So this is the packet lifecycle from beginning:

1. packet from host -> service, goes into OVN, snat to V4HostMasqueradeIP, send to OVN
2. OVN DNAT's to service endpoint (back to the host), and SNATs to V4JoinSubnet, sends packet back to br-ex
3. Flow on line 240 handles this packet and DNATs the packet to the host IP, sends the packet into the host with source IP V4JoinSubnet, dest IP host
4. Host replies to the packet, and sends back a packet destined to V4JoinSubnet, source dest IP host
5. Now this flow gets hit, which checks from LOCAL the dest IP is V4JoinSubnet (now we know this packet must be a reply, so we can unDNAT it, which will change the source back to V4HostMasqueradeIP, and send it back into OVN
6. OVN will have an existing CT session for this, and unSNAT (change dest from v4JoinSubnet back to V4HostMasqueradeIP) and unDNAT (change source back to service IP), then send back to br-ex
7. Packet will hit the flow on line 316, and unSNAT (change dest ip from V4HostMasqueradeIP back to host IP), and send the packet back to the host with src ip == service ip

**Comment by trozet** [comment]
closing in favor of #2055

**Pull request closed without merging**